Adaptive LM

gabriel-woodbury opened the pull request:
Added functionality to allow the LM parameter set in the input file to change during a run of SIESTA when it fails to improve convergence on several consecutive steps.

cianciosa committed:
Merge pull request #108 from ORNL-Fusion/master_dev

Master dev

gabriel-woodbury committed:
Change how the LM parameter is updated

This allows for the initially chosen LM parameter to change during a SIESTA run. This is meant to help SIESTA convergence when it is stalling.

cianciosa commented on Sources/hessian.f90:
```diff
@@ -56,7 +56,7 @@ MODULE hessian
                            mupar, mupar0, maxeigen, mindiag
       LOGICAL, PUBLIC   :: l_Compute_Hessian, l_Diagonal_Only
       LOGICAL, PUBLIC, PARAMETER :: l_Hess_sym=.FALSE.                     !<forces symmetrization of Hessian if TRUE (only implemented for Thomas, 1 proc)
-      INTEGER, PUBLIC :: HESSPASS=0
```
For new code we should add doxygen documentation for the variable. It should be put on its own line with a Doxygen comment. Also consider a more descriptive name for the counter_g variable.
 
```
      INTEGER, PUBLIC :: HESSPASS=0
!>  Counter for the number of Levenberg-Marquart parameter resets.
      INTEGER, PUBLIC :: counter_g=0
```

cianciosa commented on Sources/evolution.f90:
```diff
@@ -425,10 +425,20 @@ SUBROUTINE evolve
+		 ELSE
+			counter_g = 0
          END IF
+		! Changes end here 
```
This comment is not necessary.

cianciosa commented on Sources/evolution.f90:
```diff
@@ -425,10 +425,20 @@ SUBROUTINE evolve
      &      / (fsq_max - MAX(ftol, ftol_min))
          levmarq_param = f1*fsq_total1 + f2
-
+		! Changes are below. Using new counter_g to keep track of number of iterations
```
For new code, comments should start at column zero and not extend past 80 characters.

gabriel-woodbury committed:
Addressing comments in pull request

Making minor changes to the code addressing comments left in the initial pull request.

gabriel-woodbury commented:
I made some changes to the code to address your comments.

cianciosa commented on Sources/evolution.f90:
```diff
@@ -425,10 +425,20 @@ SUBROUTINE evolve
-            levmarq_param = levmarq_param0
+            IF (failed_min_count .eq. 3) THEN
+				levmarq_param0 = levmarq_param0*3
+				failed_min_count =0
```
Should be space between the `=0`.
```
failed_min_count = 0
```

cianciosa commented:
Additionally you should add yourself as an author since you contributed code.

https://github.com/ORNL-Fusion/SIESTA/blob/3ff17fee5d84ef06cbedbfbede099a2f612769da/Sources/siesta.f90#L15

gabriel-woodbury committed:
Update siesta.f90

gabriel-woodbury committed:
Update evolution.f90

Minor edits to adaptive algorithm. Needs functionality to stop code when max LM parameter has been reached and fsq does not decrease for a certain number of iterations.

gabriel-woodbury commented:
I added myself as an author and added a functionality to limit the maximum LM parameter. The next thing that I think should be added is a check on if the LM parameter has reached or exceeded this max value. Then if the LM parameter stays at this value without improving the fsq for a certain number of iterations to have the code to stop as if it had reached its iteration limit since there will be no benefit in running the code any longer. I am unsure how to do this last bit.

gabriel-woodbury committed:
Update evolution.f90

Added changing of LM parameter to case when fsq is greater than fsq_max
